### PR TITLE
Flink: add util class to generate test data with extensive coverage d…

### PR DESCRIPTION
**Pull request by stevenzwu**
…ifferent field types: from primitives to complex nested types

### CODE REVIEW

**Comment by stevenzwu** [comment]
We had this util class internally for testing Avro GenericRecord to Flink RowData converter. It can be useful for writing unit test for the `StructRowData` class from PR #6222.

**Comment by stevenzwu** [comment]
@hililiwei please help review this PR

**Comment by hililiwei** [comment]
Could we have a class that generates random data according to the schema passed in by user?

**Comment by chenjunjiedada** [comment]
In flink-1.13 test source, we have `RandomRowData` class that should work.

**Comment by stevenzwu** [comment]
@hililiwei @chenjunjiedada don't want to use random data. The purpose of this util class is to generate Flink `RowData`, Iceberg `GenericRecord`, and Avro `GenericRecord` (internal code) with the same schema and values. Then we can compare the expected and actual values after applying converter. With random values, we can not do `assertEquals`. The goal the util class is to cover different field types (primitives and complex nested). Random values are really important for that purpose.

```
  protected void testConverter(DataGenerator dataGenerator) throws Exception {
    RowDataToAvroGenericRecordConverter converter =
        RowDataToAvroGenericRecordConverter.fromAvroSchema(dataGenerator.avroSchema());
    GenericRecord expected = dataGenerator.generateAvroGenericRecord();
    GenericRecord actual = converter.apply(dataGenerator.generateFlinkRowData());
    Assert.assertEquals(expected, actual);
  }
```

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/DataGenerators.java [hunk 1]
```diff
@@ -0,0 +1,746 @@
...
+    public GenericRecord generateIcebergGenericRecord() {
+      GenericRecord genericRecord = GenericRecord.create(icebergSchema);
+      genericRecord.setField("partition_field", "partition_value");
+      genericRecord.setField("array_of_int", Arrays.asList(1, 2, 3));
+      return genericRecord;
+    }
+
+    @Override
+    public RowData generateFlinkRowData() {
+      Integer[] arr = {1, 2, 3};
```
Comment (hililiwei) [inline flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/DataGenerators.java:399]: `arr` seems to be class var, `generateIcebergGenericRecord` can also use it?

**Review by hililiwei — COMMENTED** [review]
Cool, looks good to me. Leaving a few comments.

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/DataGenerators.java [hunk 2]
```diff
@@ -0,0 +1,746 @@
...
+    private static final int DAYS_BTW_EPOC_AND_20220110 =
+        Days.daysBetween(JODA_DATETIME_EPOC, JODA_DATETIME_20220110).getDays();
+    private static final int HOUR_8_IN_MILLI = (int) TimeUnit.HOURS.toMillis(8);
+
+    private static final LocalDate JAVA_LOCAL_DATE_20220110 = LocalDate.of(2022, 1, 10);
+    private static final LocalTime JAVA_LOCAL_TIME_HOUR8 = LocalTime.of(8, 0);
+    private static final OffsetDateTime JAVA_OFFSET_DATE_TIME_20220110 =
+        OffsetDateTime.of(2022, 1, 10, 0, 0, 0, 0, ZoneOffset.UTC);
+    private static final LocalDateTime JAVA_LOCAL_DATE_TIME_20220110 =
+        LocalDateTime.of(2022, 1, 10, 0, 0, 0);
```
[the remark this answers is not in the document]
Comment (stevenzwu) [reply flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/DataGenerators.java:68]: not sure if it is necessary. will stay put.

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/DataGenerators.java [hunk 3]
```diff
@@ -0,0 +1,746 @@
...
+      }
+      genericRecord.setField("uuid_field", uuidBytes);
+
+      byte[] binaryBytes = new byte[7];
+      for (int i = 0; i < 7; ++i) {
+        binaryBytes[i] = (byte) i;
+      }
+      genericRecord.setField("binary_field", ByteBuffer.wrap(binaryBytes));
+
+      String decimalString = "-1.50";
```
[the remark this answers is not in the document]
Comment (stevenzwu) [reply flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/DataGenerators.java:129]: not sure if  it is necessary. will probably stay put on this one.

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/DataGenerators.java [hunk 4]
```diff
@@ -0,0 +1,746 @@
...
+          StringData.fromString("str"),
+          DAYS_BTW_EPOC_AND_20220110,
+          HOUR_8_IN_MILLI,
+          // Although Avro logical type for timestamp fields are in micro seconds,
+          // AvroToRowDataConverters only looks for long value in milliseconds.
+          TimestampData.fromEpochMillis(JODA_DATETIME_20220110.getMillis()),
+          TimestampData.fromEpochMillis(JODA_DATETIME_20220110.getMillis()),
+          uuidBytes,
+          binaryBytes,
+          DecimalData.fromBigDecimal(new BigDecimal("-1.50"), 9, 2));
```
[the remark this answers is not in the document]
Comment (stevenzwu) [reply flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/DataGenerators.java:164]: agree. will update

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/DataGenerators.java (hunk 1 above)
Comment (stevenzwu) [reply flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/DataGenerators.java:399]: `generateIcebergGenericRecord` needs a list as `Arrays.asList(1, 2, 3)` . so it is not reusable.

**Review by hililiwei — APPROVED** [review]

**Review by pvary — APPROVED** [review]
+1 LGTM

We might want to add some more implementations:
- Null values
- Why did we leave out Fixed?
- Edge cases, like year differences based on TZ, or non valid timestamps based on TZ
- ...

But we can add those later if needed.
The IF and the basic implementations are ok.

**Comment by stevenzwu** [comment]
thanks @hililiwei and @pvary for reviewing. will follow up on Peter's comment with a separate PR

**Comment by stevenzwu** [comment]
@pvary I missed the FixedType because of this code snippet in Iceberg

```
  private static final ImmutableMap<String, PrimitiveType> TYPES =
      ImmutableMap.<String, PrimitiveType>builder()
          .put(BooleanType.get().toString(), BooleanType.get())
          .put(IntegerType.get().toString(), IntegerType.get())
          .put(LongType.get().toString(), LongType.get())
          .put(FloatType.get().toString(), FloatType.get())
          .put(DoubleType.get().toString(), DoubleType.get())
          .put(DateType.get().toString(), DateType.get())
          .put(TimeType.get().toString(), TimeType.get())
          .put(TimestampType.withZone().toString(), TimestampType.withZone())
          .put(TimestampType.withoutZone().toString(), TimestampType.withoutZone())
          .put(StringType.get().toString(), StringType.get())
          .put(UUIDType.get().toString(), UUIDType.get())
          .put(BinaryType.get().toString(), BinaryType.get())
          .buildOrThrow();
```

**Pull request merged**